### PR TITLE
[V1][Sampler] Faster top-k only implementation

### PR DESCRIPTION
**Pull request by njhill**
When there's top-k in the batch but no top-p.

For 128k vocab, 1024 batch size, 500 ops on A100, where max top k is 10:

Before: 11.571 sec
After: 2.136 sec

### CODE REVIEW

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 1]
```diff
@@ -138,8 +138,23 @@ def apply_top_k_top_p(
...
+
+        no_top_k_mask = k == logits.shape[1]
+        # Set non-top-k rows to 1 so that we can gather.
+        k = k.masked_fill(no_top_k_mask, 1).unsqueeze(1)
+        max_top_k = k.max()
+        k.sub_(1)
+        top_k_mask = logits.topk(max_top_k, dim=1).values.gather(1, k)
+        # Handle non-topk rows.
+        top_k_mask.masked_fill_(no_top_k_mask.unsqueeze(1), -float("inf"))
+        logits.masked_fill_(logits < top_k_mask, -float("inf"))
```
Comment (njhill) [inline vllm/v1/sample/ops/topk_topp_sampler.py:155]: This is modifying logits in-place. It's slightly slower if changed to out-of-place but still much faster than the other impl.

**Review by NickLucche — APPROVED** [review]
Tested on TPU this won't work out of the box due to some broadcasting issue.

**Comment by njhill** [comment]
@NickLucche that's strange. Which op has that issue?

**Comment by NickLucche** [comment]
Not too surprising, torch xla has more constraining rules on broadcasting.
This is the first error I have encountered
```
F0325 16:28:32.957930 1304047 debug_macros.h:21] Non-OK-status: status.status()
Status: INVALID_ARGUMENT: Input dimension should be either 1 or equal to the output dimension it is broadcasting into; the 0th operand dimension is 4, the 0th output dimension is 1.
*** Begin stack trace ***
	tsl::CurrentStackTrace[abi:cxx11]()
	xla::Shape const* ConsumeValue<xla::Shape const*>(absl::lts_20230802::StatusOr<xla::Shape const*>&&)
	torch_xla::ShapeHelper::ShapeOfXlaOp(xla::XlaOp)
	torch_xla::InferOutputShape(absl::lts_20230802::Span<xla::Shape const>, std::function<xla::XlaOp (absl::lts_20230802::Span<xla::XlaOp const>)> const&)
	
	
	torch_xla::XlaNode::GetOpShape(std::function<xla::Shape ()> const&) const
	torch_xla::XlaNode::XlaNode(torch::lazy::OpKind, c10::ArrayRef<torch::lazy::Value>, std::function<xla::Shape ()> const&, unsigned long, torch::lazy::hash_t)
	torch_xla::Gather::Gather(torch::lazy::Value const&, long, torch::lazy::Value const&)
	std::shared_ptr<torch::lazy::Node> torch_xla::MakeNode<torch_xla::Gather, torch::lazy::Value, long&, torch::lazy::Value>(torch::lazy::Value&&, long&, torch::lazy::Value&&)
	torch_xla::tensor_methods::gather(c10::intrusive_ptr<torch_xla::XLATensor, c10::detail::intrusive_target_default_null_type<torch_xla::XLATensor> > const&, long, c10::intrusive_ptr<torch_xla::XLATensor, c10::detail::intrusive_target_default_null_type<torch_xla::XLATensor> > const&)
	torch_xla::XLANativeFunctions::gather(at::Tensor const&, long, at::Tensor const&, bool)
	
	
	at::_ops::gather::redispatch(c10::DispatchKeySet, at::Tensor const&, long, at::Tensor const&, bool)
	
	
	at::_ops::gather::call(at::Tensor const&, long, at::Tensor const&, bool)

```
on the `.gather` op. I expanded `k` but then ran into another issue.

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 2]
```diff
@@ -138,8 +138,23 @@ def apply_top_k_top_p(
...
+        if k is None:
+            return logits
+
+        # Avoid sorting vocab for top-k only case.
+
+        no_top_k_mask = k == logits.shape[1]
+        # Set non-top-k rows to 1 so that we can gather.
+        k = k.masked_fill(no_top_k_mask, 1).unsqueeze(1)
+        max_top_k = k.max()
+        k.sub_(1)
```
Comment (WoosukKwon) [inline vllm/v1/sample/ops/topk_topp_sampler.py:151]: A dumb question: Why -1 here?

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/ops/topk_topp_sampler.py:155]: Then, should we have a comment on this in the docstring and the caller of this function?

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 2 above)
Comment (njhill) [reply vllm/v1/sample/ops/topk_topp_sampler.py:151]: Your questions are never dumb @WoosukKwon :)

It's just converting to 0-based index for the gather. `top_k_mask.shape[1]` will be `max_top_k`, so the request with max top k will be out-of-bounds for example.

And we're masking logits that are less than these selected ones so I think it works out (e.g. for k=1 we want to select the 0th element of second dimension and then exclude all with lower logit value than that...)

Comment (WoosukKwon) [reply vllm/v1/sample/ops/topk_topp_sampler.py:151]: Got it. Thanks for the explanation.
Could you also put it in the code so that anyone can easily understand?

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/sample/ops/topk_topp_sampler.py:157]: @njhill ^^

File: vllm/v1/sample/ops/topk_topp_sampler.py [hunk 3]
```diff
@@ -138,8 +138,25 @@ def apply_top_k_top_p(
 
     This function sorts the logits tensor, which can be slow for large batches.
     """
-    if k is None and p is None:
+    if p is None:
+        if k is None:
```
Comment (WoosukKwon) [inline vllm/v1/sample/ops/topk_topp_sampler.py:142]: Do we have a unit test checking the correctness of this?

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 1 above)
Comment (njhill) [reply vllm/v1/sample/ops/topk_topp_sampler.py:170]: Sorry, missed the comment originally. I've added some comments about this now. Some of the other methods do already update the logits in-place (temperature, min_p, etc.) so I do think it's fine/expected here.

File: vllm/v1/sample/ops/topk_topp_sampler.py (hunk 3 above)
Comment (njhill) [reply vllm/v1/sample/ops/topk_topp_sampler.py:155]: We should really have blanket coverage for this kind of thing, including different combinations of parameters (i.e. top-k with/without top-p etc.). I'm not sure whether we do though. I will check and add a unit test to compare the two impls.

**Review by NickLucche — CHANGES_REQUESTED** [review]
I tested this version again today and it's working on TPU too, nice one @njhill thanks!
I was wondering could we still factor-out this topk opt into its own function so I can call it from TPU side?
We agreed with @WoosukKwon to try and keep things separated, I'd like to keep `forward_tpu` around.

**Comment by NickLucche** [comment]
Something like https://github.com/vllm-project/vllm/pull/15489/commits/a5bf8492360640a645886de474c2cdb42003ecbd#diff-6047245d864bf5fd68b5b947b735beca94723bad40d20bfc0803d9b3eea5c1edR121-R136. 
Wdyt? Of course I'd wait for this PR to land and then rebase, I've shamelessly just copy-pasted your code there.

**Comment by njhill** [comment]
Thanks @NickLucche, I've split into separate function. And @WoosukKwon I've added a correctness test.

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for addressing my comments.

**Comment by hyeygit** [comment]
@njhill really neat idea to threshold the logits! However I think one corner case where this would break is if there are duplicate elements in the logit that equal the cut off value (i.e. `top_p_mask`). For example, given an input of `[1, 2, 2, 2, 3]` and `k=3`, the current `apply_top_k_only` would return `[-inf, 2, 2, 2, 3]` while the correct result should be `[-inf, -inf, 2, 2, 3]`.

In #15736 I use a similar thresholding logic for top-p, but introduced a small random perturbation to break the ties. Maybe the same idea can be used here for top-k as well.
